---
title: "Exercise - Proteolytic cleavage; Beta-Lactoglobulin"
author: "Marc A.T. Teunis"
date: '`r Sys.Date()`'
output: 
  rmdformats::downcute:
    self_contained: TRUE
    number_sections: FALSE
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      error = FALSE,
                      results = 'show',
                      fig.show = 'show')
``` 

# Setting the project root
We need an object for the root of the project for purposes yet to explain.
```{r, echo=TRUE, eval=TRUE, results='markup'}
#if (!require("rprojroot")) install.packages("rprojroot")
#library("rprojroot")
#root <- find_root_file(criterion = is_rstudio_project)
#root
```

# Introduction
We use the protein Beta-lactoglobulin as an example to show an _in-slico_ chemical analysis workflow in R/Bioconductor.

Here we will perform an enzymatic digestion on the native beta-lactoglobulin (`BLG`) protein of different species. We will use the Cow, Goat and the Sheep sequence to explore the theory behind mass-spectrometery. The amino acid sequence for the Beta-lac (`BLG`) gene for different organisms (Cow, Sheep & Goat) will be used.

We also practice a little with alignment in this exercise.

The name (`external_gene_name`) of the beta-lac encoding gene is `BLG`. Synonyms are: `PAEP`,  `LGB` or `Beta-LG`

# Packages
Load the following packages in your session:
```{r, echo=TRUE}

library(biomaRt)
library(seqinr)
library(rentrez)
library(GEOquery)
library(Biostrings)
library(DECIPHER)
library(cleaver)
library(OrgMassSpecR) 
library(protr)
library(tidyverse)
library(here)
```

# Data

Obtain the amino-acid sequences for the Beta-Lactoglobuline protein (`BLG`) from Goat (_C. hircus_), Sheep (_O. aries_) and Domestic Cow (_B. taurus_).
See code chunks below.

## Retrieve amino acid sequence of `BLG` protein for Goat  
Because the Goat sequence is not available via a Biomart search we download this sequence directly from Uniprot.

The following chunk downloads the amino acid sequence for the `BLG` protein from here:  
https://www.uniprot.org/uniprot/P02756.fasta
And put the downloaded `.fasta` file in the `./data` folder. 

```{r, answer_1a_1, echo=TRUE}
## download Goat aa sequence
## dir.create(here("data")) 
url <- c("https://www.uniprot.org/uniprot/P02756.fasta")
download.file(url = url, destfile = here("data", "chircus_blg_aminoacids.fasta"))
```

## Convert to Amino Acids
```{r}
chircus_blg_aa <- Biostrings::readAAStringSet(filepath = here("data", "chircus_blg_aminoacids.fasta"))
names(chircus_blg_aa) <- c("chircus")
chircus_blg_aa
```

Alternatively, you can also use the `{protr}` package to download sequences directly from UniProt:

You need to know the specific UniProt identifier. For Goat `BLG` this is `B2YKY6`. The result is a list of amino acid sequence(s). You can convert this list to a more convenient object with the code below, using `biostrings::AAString()` or `biostrings::AAStringSet()`. You need to un-list the list first for this to work.
```{r, demo_protr, echo=TRUE}
protr::getUniProt("P02756") %>% 
  unlist %>%
  AAStringSet()
```

## Retrieve aa sequence of `BLG` protein for the Cow (`_B. taurus_`).
The BLG protein has multiple synonyms. The code below uses an `lapply` loop to test for each known synonym: `BLG`, `PAEP` `Beta-LG` and `LGB` 
```{r, demo_cow_seq, echo=TRUE}
## specify the 'market place' to use
mart <- useMart(biomart = "ensembl")
## specify from which 'dataset', in this case from Domestic Cow
btaurus_mart <- useDataset(
  dataset = "btaurus_gene_ensembl", 
  mart = mart)

## there are several names for the same protein
beta_lac_gene_name <- c("BLG", "PAEP", "LGB", "Beta-LG")

attributes <- listAttributes(mart = btaurus_mart)

loop_gene_names <- function(x, mart){
  results <- getBM(
    attributes = c("ensembl_gene_id",        ## attributes are output
                   "external_gene_name",
                    "coding"), 
    filters = "external_gene_name",          ## filters in input type
    values = x, 
    mart = mart) %>%
    as_tibble() 
    # print()
  return(results)
}

## "PAEP" yields a result, using that
query <- lapply(as.list(beta_lac_gene_name), 
                loop_gene_names, 
                mart = btaurus_mart)
query
names(query) <- beta_lac_gene_name

## isolating the result in a dataframe
query_btaurus <- query[["PAEP"]] 

btaurus_blg_aa <- DNAStringSet(query_btaurus$coding) %>% Biostrings::translate()
btaurus_blg_aa@ranges@NAMES <- c("btaurus", "btaurus", "btaurus")
btaurus_blg_aa
```

## Retrieve aa sequence for Sheep
Now do the same for the Sheep `BLG` protein sequence. Use the code example above.
```{r, answer_1a}
mart <- useMart(biomart = "ensembl")
oaries_mart <- useDataset(dataset = "oaries_gene_ensembl", 
                        mart=mart)

#datasets <- listDatasets(mart) %>% as_tibble()
attributes <- listAttributes(mart = oaries_mart)
#filters <- listFilters(chircus_mart)

## using loop again
query_sheep <- lapply(as.list(beta_lac_gene_name), loop_gene_names, mart = oaries_mart)
query_sheep
names(query_sheep) <- beta_lac_gene_name
query_sheep <- query_sheep[["PAEP"]]
query_sheep

oaries_blg_aa <- DNAStringSet(query_sheep$coding) %>% Biostrings::translate()
names(oaries_blg_aa) <- "oaries"
oaries_blg_aa

```

Combine the three seperate amino acid sequences encoding the `BLG` protein from Cow, Goat and Sheet to one `AAStringSet`. Provide appropriate names for each sequence by setting names using `names(...) <- c(...)`. 

How long is each protein(fragment)?

```{r, answer_1b}


all_aa_trim <- c(btaurus_blg_aa, 
                 oaries_blg_aa, 
                 chircus_blg_aa)
all_aa_trim
```

## Exercise signal peptide trim and compare to Robert Jan
```{r}
all_aa_trim_trim <- AAStringSet(all_aa_trim, start = 17)
all_aa_trim_trim

rjve_seq <- "LIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEELKPTPEGDLEILLQKWENDECAQKKIIAEKTKIPAVFKIDALNENKVLVLDTDYKKYLLVCMENSAEPEQSLVCQCLVRTPEVDDEALEKFDKALKALPMHIRLSFNPTQLEEQCHI"

rjve_seq <- AAStringSet(rjve_seq)
rjve_seq@ranges@NAMES <- "RJVE_seq"

new_sequences <- c(rjve_seq, all_aa_trim_trim)

alignment_blg <- AlignSeqs(new_sequences[1:3])
BrowseSeqs(alignment_blg)

alignment_blg

```

# Multiple sequence alignment
**2A**
Align the amino acid sequences for the `BLG` protein from Cow, Sheep and Goat. What can you conclude from this alignment?
```{r, answer_2a}
alignment_blg <- AlignSeqs(all_aa_trim_trim)
BrowseSeqs(alignment_blg)
```

Align the sequences for `BLG` protein from Cow and Sheep, what is you conclusion about the homology 
```{r, answer_2b_1, results='show', fig.show='show'}
alignment_blg <- AlignSeqs(all_aa_trim_trim[c(1:4)])
BrowseSeqs(alignment_blg)
```

Now do the same for the sequence from Cow and Goat
Which animal is most related to the Cow, on the basis of this alignment? Can you think of a way to illustrate this further?
```{r, answer_2c}
alignment_blg <- AlignSeqs(all_aa_trim[c(1, 5)])
BrowseSeqs(alignment_blg)
```

## Simularity scores
Calculate the percentage simularity on the basis of the primary amino acid sequence between the `BLG` proteins. Use the _Bos taurus_ sequence as a reference sequence (`pattern`).

**TIPS**

 - use the `biostrings::pairwiseAlignment()` function
 - use the `biostrings::pid()` function to get the simularity score, use the setting in this function `type = 'PID1'`. Look at the help file of this function to see what this setting does.
 
```{r, answer_2d}
all_aa_trim

btaurus_vs_oaries <- pairwiseAlignment(
  pattern = all_aa_trim[1], 
  subject = all_aa_trim[2]
  )
btaurus_vs_chircus <- pairwiseAlignment(
  pattern = all_aa_trim[1], 
  subject = all_aa_trim[3]
  )
oaries_vs_chircus <- pairwiseAlignment(
  pattern = all_aa_trim[2], subject = all_aa_trim[3]
  )

?pid

pid_string <- "PID1"

# percentage simularity for muliple alignments 
pid(btaurus_vs_oaries, type = pid_string)
pid(btaurus_vs_chircus, type = pid_string)
pid(oaries_vs_chircus, type = pid_string)


```

# Proteolytic digestion

Review the vignette from the `{OrgMassSpecR}` package. 
Try predicting the fragments by proteolytic digestion 
with the trypsin enzyme, for the _Bos taurus_ BLG protein. 

**TIPS:**

 - Look at the vignette for the `{OrgMassSpecR}` package, it has some really good examples.
 - set the `enzyme` option to `trypsin`
 - Review the code below from the presentation "Mass Spectometry": recycle the code to find the solution for 3A

```{r, demo_digestion, echo=TRUE}
browseVignettes(package = "OrgMassSpecR")

seq <- all_aa_trim %>%
  as.list()

## remove * from sequences
#seq <- map(seq, str_remove_all, pattern = "\\*")
seq <- modified_aa_list <- lapply(seq, function(x) gsub("\\*$", "", x))  

btaurus_blg_trypsin <- OrgMassSpecR::Digest(
  sequence = seq_modified[[1]], enzyme ="trypsin"
  ) %>% 
  as_tibble() %>%
  mutate(width = nchar(peptide)) %>%
  arrange(width) %>% 
  print()
```

## Write a function that takes a sequence and returns a nice table
```{r}
## a single sequence as input:
digest_sequence <- function(peptide, enzyme){
  
  digestion <- OrgMassSpecR::Digest(peptide, enzyme = enzyme) %>% 
     as_tibble() %>%
     mutate(width = nchar(peptide)) %>%
     arrange(width) 
  
  return(digestion)
  
}

digest_sequence(seq[[1]], enzyme = "trypsin")

## loop over each sequence
all_predictions <- map(seq, digest_sequence, enzyme = "trypsin")

all_predictions[[1]]
all_predictions[[2]]
all_predictions[[3]]
```

## Let's examine the Cow a bit further - We only examine the first sequence of the AAStringSet
```{r, answer_3a}

## store as xlsx
readr::write_csv(
  btaurus_blg_trypsin,
  file = here::here("data",
                   "btaurus_blg_trypsin.csv"))

## get the cleavage ranges
ranges <- cleavageRanges(all_aa_trim$btaurus, enzym = "trypsin") 

tibble(width = ranges@width, 
       start = ranges@start) %>%
  arrange(width)

## get only cleavage sites
sites <- cleavageSites(all_aa_trim[1], enzym = "trypsin")
sites

```

**3B**
Recycling the code from question 3A, try predicting the peptide-fragments from the Goat and the Sheep. Combine these in a IRanges object, with appropriate names (use for example Species and width). Use `lapply` to achieve your goal. 

# Mass spectrometric analysis  

## Get all structure formulas for each peptide in all digestions
```{r}
formula <- OrgMassSpecR::ConvertPeptide(
  btaurus_blg_trypsin$peptide[1])

formula

MonoisotopicMass(formula = ConvertPeptide(btaurus_blg_trypsin$peptide[1]))
MolecularWeight(formula = ConvertPeptide(btaurus_blg_trypsin$peptide[1]))

MonoisotopicMass(formula = ConvertPeptide(btaurus_blg_trypsin$peptide[19]))
MolecularWeight(formula = ConvertPeptide(btaurus_blg_trypsin$peptide[19]))
## Compare to results of Robert Jan on slide 21

as_tibble(formula)

# btaurus_blg_trypsin$fragment

get_formula <- function(peptide){
  
  formula <- OrgMassSpecR::ConvertPeptide(peptide)
  df <- formula %>% as_tibble()

    return(df)

}
  
get_formula(peptide = btaurus_blg_trypsin$peptide[1])  
  

peptide_list <- as.list(btaurus_blg_trypsin$peptide)

## safe loop with error catching
formulae <- map(peptide_list, safely(get_formula))  
formulae <- purrr::transpose(formulae)
formulae$result %>% dplyr::bind_rows()
formulae$error
```

## Get all Monoisotopic masses
```{r}
get_monoisotopic_mass <- function(peptide){
  
  formula <- OrgMassSpecR::MonoisotopicMass(
    ConvertPeptide(peptide))
  df <- formula %>% as_tibble()
  df$peptide <- peptide
    return(df)

}

get_monoisotopic_mass(peptide = btaurus_blg_trypsin$peptide[1])

map(peptide_list, get_monoisotopic_mass) %>%
  dplyr::bind_rows() %>%
  rename(monoisotopic_mass = value)
  

```

## Isotopic Distribution
```{r}
id <- IsotopicDistribution(formula = formula) %>%
  as_tibble() %>%
  print()
```

## Plot Isotopic Distribution MZ - Intensity
```{r}
ggplot(data = id, aes(x = mz, y = intensity)) +
  geom_point() +
  geom_segment(aes(xend = mz, yend=-Inf)) +
  ggtitle("Fragment 1, Bos taurus")

ggplot(data = id, aes(x = mz, y = log10(intensity))) +
  geom_point() +
  geom_segment(aes(xend = mz, yend=-Inf))

```

## Longest fragment
```{r}

IsotopicDistribution(formula = get_formula(peptide = btaurus_blg_trypsin$peptide[19])) %>%
  as_tibble() %>%
    ggplot(aes(x = mz, y = intensity)) +
      geom_point() +
      geom_segment(aes(xend = mz, yend=-Inf)) +
      ggtitle("Fragment 19, Bos taurus")


```

## Identification
```{r}
### Next step is identification
```
